lib/types/hostnametrie.go

```diff
@@ -135,42 +138,52 @@ func (t *HostnameTrie) insert(s string) error {
 		return err
 	}
 
-	return t.childInsert(s)
+	return t.trieNode.insert(s)
+}
+
+// Contains returns whether s matches a pattern in the HostnameTrie
+// along with the matching pattern, if one was found.
+func (t *HostnameTrie) Contains(s string) (matchedPattern string, matchFound bool) {
+	return t.trieNode.contains(s)
 }
 
-func (t *HostnameTrie) childInsert(s string) error {
+type trieNode struct {
+	isLeaf   bool
+	children map[rune]*trieNode
+}
+
+func (t *trieNode) insert(s string) error {
 	if len(s) == 0 {
+		t.isLeaf = true
```

Comment:
Shouldn't this become ... false if it goes beyond this ?

Reply:
I don't understand - the next line is `return nil`, so it can't go beyond this :confused:

Reply:
I meant that if on sequential call to insert on a node that is already a leaf ... we won't unleaf it ... which I now understand is exactly what this is fixing (but I needed to read the whole code ...). 

I am somewhat of the opinion that if there is "*valid.test" and "test.valid,test" we should probably not add the second one to tree ... and/or give out a warning/error that there is a more general glob already. 

But this can be done as an optimization in later PR and I am not particularly opinioanted about it ...